packages/Search/src/details/DTK_DetailsTreeConstruction_decl.hpp

```diff
@@ -95,6 +101,169 @@ struct TreeConstruction
+}
+
+template <typename DeviceType>
+template <typename ConstViewType>
+inline void TreeConstruction<DeviceType>::assignMortonCodes(
+    ConstViewType primitives,
+    Kokkos::View<unsigned int *, DeviceType> morton_codes,
+    Box const &scene_bounding_box )
+{
+    auto const n = primitives.extent( 0 );
+    DTK_REQUIRE( morton_codes.extent( 0 ) == n );
+
+    using Tag = typename Tag<typename decltype(
+        primitives )::traits::non_const_value_type>::type;
+    assignMortonCodesDispatch( Tag{}, primitives, morton_codes,
+                               scene_bounding_box );
+}
+
+template <typename Primitives, typename Indices, typename Nodes>
+inline void initializeLeafNodesDispatch( BoxTag, Primitives primitives,
+                                         Indices permutation_indices,
+                                         Nodes leaf_nodes )
+{
+    using ExecutionSpace =
+        typename decltype( primitives )::traits::execution_space;
+    auto const n = leaf_nodes.extent( 0 );
+    Kokkos::parallel_for(
+        DTK_MARK_REGION( "initialize_leaf_nodes" ),
+        Kokkos::RangePolicy<ExecutionSpace>( 0, n ), KOKKOS_LAMBDA( int i ) {
+            leaf_nodes( i ) = {
+                {nullptr, reinterpret_cast<Node *>( permutation_indices( i ) )},
+                primitives( permutation_indices( i ) )};
+        } );
+    Kokkos::fence();
+}
+
+template <typename Primitives, typename Indices, typename Nodes>
+inline void initializeLeafNodesDispatch( PointTag, Primitives primitives,
+                                         Indices permutation_indices,
+                                         Nodes leaf_nodes )
+{
+    using ExecutionSpace =
+        typename decltype( primitives )::traits::execution_space;
+    auto const n = leaf_nodes.extent( 0 );
+    Kokkos::parallel_for(
+        DTK_MARK_REGION( "initialize_leaf_nodes" ),
+        Kokkos::RangePolicy<ExecutionSpace>( 0, n ), KOKKOS_LAMBDA( int i ) {
+            leaf_nodes( i ) = {
+                {nullptr, reinterpret_cast<Node *>( permutation_indices( i ) )},
+                {primitives( permutation_indices( i ) ),
+                 primitives( permutation_indices( i ) )}};
+        } );
+    Kokkos::fence();
+}
+
+template <typename DeviceType>
+template <typename ConstViewType>
+inline void TreeConstruction<DeviceType>::initializeLeafNodes(
+    ConstViewType primitives,
+    Kokkos::View<size_t const *, DeviceType> permutation_indices,
+    Kokkos::View<Node *, DeviceType> leaf_nodes )
+{
+    auto const n = leaf_nodes.extent( 0 );
+    DTK_REQUIRE( permutation_indices.extent( 0 ) == n );
+    DTK_REQUIRE( primitives.extent( 0 ) == n );
+
+    static_assert( sizeof( typename decltype(
+                       permutation_indices )::value_type ) == sizeof( Node * ),
+                   "Encoding leaf index in pointer to child is not safe if the "
+                   "index and pointer types do not have the same size" );
+
+    using Tag = typename Tag<typename decltype(
```

Comment:
Don't need `decltype` here, just use `ConstViewType`. Or did you like this better?

Reply:
No I did it on purpose because I kept on changing the template argument name.
This will change soon anyway, I am about to introduce traits to interact with the input argument pass to the BVH constructor.  This will allow building from a `View<double*[3]` for instance.